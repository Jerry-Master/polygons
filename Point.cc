#include "Point.h"

// Constructor
Point& Point::operator= (const Point& P){
    x = P.X(); y = P.Y();
    return *this;
}
Point::Point (double x2, double y2) : x(x2), y(y2) {}

// Coordinates
double Point::X () const {
    return x;
}
double Point::Y () const {
    return y;
}

// Important values
// Quadrance, just to avoid square roots
double Point::quadrance() const {
    return x*x + y*y;
} 
// To the x-axis
double Point::slope() const {
    if (abs(x) < EPSILON) return INFINITY;
    return y/x;
}
// (first goes the x and then the y) 1: ++, 2: -+, 3: --, 4: +-, 5: axis
int Point::quadrant() const {
    if (x > 0 and y > 0) return 1;
    if (x < 0 and y > 0) return 2;
    if (x < 0 and y < 0) return 3;
    if (x > 0 and y < 0) return 4;
    // Dealing with precision issues around the axis
    if (abs(x) < EPSILON or abs(y) > EPSILON) return 5;
    else assert(false);
} 
double Point::distance(const Point& P) const{
    return sqrt((P - *this).quadrance());
}

// Add, substract, cross-product, dot product, clockwise orientation
Point& Point::operator+= (const Point& P){
    x += P.X(); y += P.Y();
    return *this;
}
Point Point::operator+ (const Point& P) const{
    Point _aux = *this;
    _aux += P;
    return _aux;
}
Point& Point::operator-= (const Point& P){
    x -= P.X(); y -= P.Y();
    return *this;
}
Point Point::operator- (const Point& P) const{
    Point _aux = *this;
    _aux -= P;
    return _aux;
}
Point Point::operator- () const{
    Point _aux(-x, -y);
    return _aux;
}
// Multiplication by a scalar
Point& Point::operator*= (double lambda){
    x *= lambda; y *= lambda;
    return *this;
}
// Cross-product
double Point::operator^ (const Point& P) const{
    return x * P.Y() - y * P.X();
}
// Dot-product
double Point::operator* (const Point& P) const{
    return x * P.X() + y * P.Y();
}
bool Point::operator== (const Point& P) const{
    return x == P.X() and y == P.Y();
}
bool Point::clockwise (const Point& P1, const Point& P2) const{
    return (P1 ^ P2) >= 0;
}
bool Point::clockwise (const Point& P) const{
    return clockwise(*this, P) > 0;
}
