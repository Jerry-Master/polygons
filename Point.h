#ifndef _POINT
#define _POINT
#define EPSILON 1e-12 

#include <cmath>
#include <cassert>

/**
This class provides all the useful functions to work with points. 
We can also identify them as vectors taking the origin as reference,
therefore all the useful operations with vectors are also implemented.

We can consider the cost of its functions as O(1)
*/

class Point{
    public:
        // Constructor
        Point& operator= (const Point& P);
        Point (double x = 0, double y = 0);

        // Coordinates
        double X () const;
        double Y () const;
        
        // Important values
        double quadrance() const; // Quadrance, just to avoid square roots
        double slope() const; // To the x-axis
        // (first goes the x and then the y) 1: ++, 2: -+, 3: --, 4: +-, 5: axis
        // To distinguish the different slopes
        int quadrant() const; 
        // Distance from this to P
        double distance(const Point& P) const;

        // Add, substract, cross-product, dot product, clockwise orientation
        Point& operator+= (const Point& P);
        Point operator+ (const Point& P) const;
        Point& operator-= (const Point& P);
        Point operator- (const Point& P) const;
        Point operator- () const;
        Point& operator*= (double lambda); // Multiplication by a scalar
        double operator^ (const Point& P) const; // Cross
        double operator* (const Point& P) const; // Dot
        bool operator== (const Point& P) const;
        // Taking the origin as reference
        // True: P1^P2 > 0
        // False: P1^P2 < 0
        // Prec: Neither P1 nor P2 are the origin
        bool clockwise (const Point& P1, const Point& P2) const;
        bool clockwise (const Point& P) const;

    private:
        double x, y;
};

#endif
