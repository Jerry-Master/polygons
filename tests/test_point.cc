#include "../Point.h"

#include <iostream>
#include <map>
using namespace std;

map<string, Point> dots;

void add_point(){
    string name;
    double x, y;
    cin >> name >> x >> y;
    dots[name] = Point(x,y);
}

void remove(){
    string name;
    cin >> name;
    dots.erase(name);
}
// Assigns the first to the second
void assign(){
    string name1, name2;
    cin >> name1 >> name2;
    dots[name2] = dots[name1];
}

void add_cum(){
    string name1, name2;
    cin >> name1 >> name2;
    dots[name1] += dots[name2];
}

void add(){
    string name1, name2;
    cin >> name1 >> name2;
    Point aux = dots[name1] + dots[name2];
    cout << aux.X() << ' ' << aux.Y() << endl;
}

void substract_cum(){
    string name1, name2;
    cin >> name1 >> name2;
    dots[name1] -= dots[name2];
}

void substract(){
    string name1, name2;
    cin >> name1 >> name2;
    Point aux = dots[name1] - dots[name2];
    cout << aux.X() << ' ' << aux.Y() << endl;
}

void scalar(){
    string name;
    double lambda;
    cin >> name >> lambda;
    dots[name] *= lambda;
    cout << dots[name].X() << " " << dots[name].Y() << endl;
}

void cross(){
    string name1, name2;
    cin >> name1 >> name2;
    double aux = dots[name1] ^ dots[name2];
    cout << aux << endl;
}

void dot(){
    string name1, name2;
    cin >> name1 >> name2;
    double aux = dots[name1] * dots[name2];
    cout << aux << endl;
}

void equal(){
    string name1, name2;
    cin >> name1 >> name2;
    cout << (dots[name1] == dots[name2] ? "yes" : "no") << endl;
}

void clockwise(){
    string name1, name2;
    cin >> name1 >> name2;
    cout << (dots[name1].clockwise(dots[name2]) ? "yes" : "no") << endl;
}

void ics(){
    string name;
    cin >> name;
    cout << dots[name].X() << endl;
}

void greek_y(){
    string name;
    cin >> name;
    cout << dots[name].Y() << endl;
}

void quad(){
    string name;
    cin >> name;
    cout << dots[name].quadrance() << endl;
}

void slope(){
    string name;
    cin >> name;
    cout << dots[name].slope() << endl;
}

void quadrant(){
    string name;
    cin >> name;
    cout << dots[name].quadrant() << endl;
}

void distance(){
    string name1, name2;
    cin >> name1 >> name2;
    cout << dots[name1].distance(dots[name2]) << endl;
}

int main(){
    cout.setf(ios::fixed);
    cout.precision(2);

    string inst;
    while (cin >> inst){
        if (inst == "add_point") add_point();
        if (inst == "remove") remove();
        if (inst == "assign") assign();

        if (inst == "+=") add_cum();
        if (inst == "+") add();
        if (inst == "-=") substract_cum();
        if (inst == "-") substract();
        if (inst == "*=") scalar();
        if (inst == "^") cross();
        if (inst == "*") dot();
        if (inst == "==") equal();

        if (inst == "clockwise") clockwise();

        if (inst == "X") ics();
        if (inst == "Y") greek_y();

        if (inst == "quad") quad();
        if (inst == "slope") slope();
        if (inst == "quadrant") quadrant();
        if (inst == "distance") distance();
    }
}
