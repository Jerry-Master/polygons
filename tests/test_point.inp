add_point A 1 5
add_point B -3 -1
add_point C 2 -2
add_point D 6 10
add_point E -10 1
add_point O 0 0
add_point F 4 0
add_point G 0 -3

^ A B
^ A C
^ A O
^ B A
^ C A

* E D
* D E

+ A C
- B D

clockwise A C
clockwise C A
clockwise B E
clockwise D E

quadrant A
quadrant B
quadrant C
quadrant D
quadrant E
quadrant O

slope A
slope B
slope C
slope D
slope E
slope F
slope G
slope O

quad A
quad B
quad C
quad D
quad E
quad F
quad G
quad O

X A
+= A B
X A
X F
-= F C
X F
Y E
+= E D
Y E

assign C B
+ O C
+ O B

== A B
== C B
== B C
== O A

*= D 0.34
*= F 0.15

distance D F