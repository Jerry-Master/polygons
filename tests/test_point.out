14.00
-12.00
0.00
-14.00
12.00
-50.00
-50.00
3.00 3.00
-9.00 -11.00
no
yes
no
yes
1
3
4
1
2
5
5.00
0.33
-1.00
1.67
-0.10
0.00
inf
inf
26.00
10.00
8.00
136.00
101.00
16.00
9.00
0.00
1.00
-2.00
4.00
2.00
1.00
11.00
2.00 -2.00
2.00 -2.00
no
yes
yes
no
2.04 3.40
0.30 0.30
3.55
